Tall labels

dmvaldman opened the pull request:
Allows for a `max_lines` parameter for text fields to truncate excessively tall labels. The current default is 5 lines. This would require documentation if implemented.

dmvaldman commented:
Feel free to rebase all these :-P

bcamper commented:
Do you expect this will merge cleanly with the articulated labels branch? :)

bcamper commented on src/styles/text/canvas_text.js:
```diff
@@ -343,3 +342,107 @@ CanvasText.font_size_re = /((?:[0-9]*\.)?[0-9]+)\s*(px|pt|em|%)?/;
+        this.height = 0;
+        this.lines = [];
+
+        this.ellipsis = '...';
```
I wonder if we should use the Unicode single-character ellipsis? I suppose font support might be iffy, so maybe not.

bcamper commented:
I haven't confirmed in-code, but it looks like there's a space being left in before the ellipsis?

![screen shot 2016-08-29 at 11 16 37 am](https://cloud.githubusercontent.com/assets/16733/18056577/6a7fabee-6dda-11e6-8fbb-76a73b3b6aa4.png)

dmvaldman commented:
I can do the merging if there's a conflict.
The space before the ellipsis is applied can be easily removed. Would you prefer that?

dmvaldman commented:
Also, nice debugging spot!

bcamper commented:
@dmvaldman re: ellipsis space, yes I think I'd prefer to not have a leading space, thanks.

bcamper commented:
Seeing some weird behavior with low `text_wrap` values, in this case `5`. Maybe anything shorter than the initial word?

![tangram- 18 4488 40 7513 -73 9693_2016-08-29_11 23 22](https://cloud.githubusercontent.com/assets/16733/18056879/ccb0b794-6ddb-11e6-9f38-35644291f229.png)

dmvaldman commented:
Is the unexpected behavior you're seeing the artificial vertical space that seems to be added when the first word is larger than the allotted spacing?

bcamper commented:
Right - it appears to be inserting an empty leading line?

bcamper commented on src/styles/text/canvas_text.js:
```diff
@@ -351,8 +351,9 @@ class MultiLine {
         this.lines = [];
 
-        this.ellipsis = '...';
+        this.ellipsis = '…';
```
Sorry, I probably shouldn't have suggested this at all. We really should research handling across fonts more in depth before changing. I suspect it may not exist in some custom fonts and could just render with unexpected results?

bcamper commented:
I can't currently see any effect from changing the `max_lines` value at run-time (e.g. in Play, though theoretically a user could programmatically update this as well or hook it up to another GUI). 

I suspect this is because `max_lines` is not part of the `TextSettings` key, so `CanvasText` is caching text strings with different `max_lines` values under the same key.

dmvaldman commented:
Should be good now. Reverted the ellipsis to basic text, fixed the spacing bug, placed the `max_lines` into the `text-settings` key generation algorithm, and refactored a bit more.

bcamper commented:
This is looking good to me.

We should add test cases for `max_lines` to our Differ unit tests. Here's current 0.9.x vs. `tall-labels` branch:

http://tangrams.github.io/differ/?1=https%3A//github.com/tangrams/differ-tests/blob/gh-pages/unit-tests/text/text01.json&2=&lib1=0.9&lib2=https%3A//precog.mapzen.com/tangrams/tangram/tall-labels/dist/tangram.debug.js&ignore1&ignore2&go

The test cases are defined here: https://github.com/tangrams/differ-tests/blob/gh-pages/unit-tests/text/text01.json

@meetar can help walk you through how to add them!

bcamper commented:
Good call on factoring out the multi-line handling too :)

bcamper commented:
Merging but please leave any notes about new differ units tests there.

dmvaldman commented:
Added some unit tests to the differ-tests repo here https://github.com/tangrams/differ-tests/commit/7e0f85ecfafcb804821bf150a4e6b0cecabb37a2

Here are some results

![1](https://cloud.githubusercontent.com/assets/796394/18143753/6b482c96-6f91-11e6-81dc-2dd5daeed096.png)

![2](https://cloud.githubusercontent.com/assets/796394/18143777/7d28745c-6f91-11e6-9610-9af3136dc65d.png)

dmvaldman commented:
The `max_lines` text attribute should also be specified in the docs for v0.10. Should I handle this or would you like to @meetar

bcamper commented:
@dmvaldman nice! I added the `needs-docs` label, and then we clear those out when we do the actual release. You could add docs for this in a branch, and then we can merge it when v0.10 is released (or just write it after we release for a small feature, I often do that too :)

bcamper commented:
(After each release, we go through all the issues tagged `needs-docs`, and then remove that label once we ensure they are documented.)
